Composer/packages/client/src/pages/publish/Publish.tsx

```diff
@@ -3,126 +3,153 @@
+      botStatus.publishTargets = publishTargets;
+      if (publishHistory[botStatus.publishTarget] && publishHistory[botStatus.publishTarget].length > 0) {
+        const history = publishHistory[botStatus.publishTarget][0];
+        botStatus.time = history.time;
+        botStatus.comment = history.comment;
+        botStatus.message = history.message;
+        botStatus.status = history.status;
+      }
+    }
+    statusList.push(botStatus);
+  });
 
+  const [botStatusList, setBotStatusList] = useState<IBotStatus[]>(statusList);
+  const [botPublishHistoryList, setBotPublishHistoryList] = useState<
+    { projectId: string; publishHistory: IStatus[] }[]
+  >(publishHistoyList);
   const [showLog, setShowLog] = useState(false);
   const [publishDialogHidden, setPublishDialogHidden] = useState(true);
   const [pullDialogHidden, setPullDialogHidden] = useState(true);
 
   // items to show in the list
-  const [thisPublishHistory, setThisPublishHistory] = useState<IStatus[]>([]);
-  const [groups, setGroups] = useState<any[]>([]);
   const [selectedVersion, setSelectedVersion] = useState<IStatus | null>(null);
-  const [dialogProps, setDialogProps] = useState({
-    title: formatMessage('Title'),
-    type: DialogType.normal,
-    children: {},
-  });
-  const [editDialogProps, setEditDialogProps] = useState({
-    title: formatMessage('Title'),
-    type: DialogType.normal,
-    children: {},
-  });
-  const [editTarget, setEditTarget] = useState<{ index: number; item: PublishTarget } | null>(null);
 
-  const isRollbackSupported = useMemo(
-    () => (target, version): boolean => {
-      if (version.id && version.status === 200 && target) {
-        const type = publishTypes?.filter((t) => t.name === target.type)[0];
-        if (type?.features?.rollback) {
-          return true;
-        }
+  const isRollbackSupported = (
+    targetName: string,
+    version: IStatus,
+    publishTargets: PublishTarget[],
+    projectId: string
+  ): boolean => {
+    const target = publishTargets.find((publishTarget) => publishTarget.name === targetName);
+    if (version.id && version.status === 200 && target) {
+      const publishTypes = botPublishTypesList.find((type) => type.projectId === projectId)?.publishTypes;
+      const type = publishTypes?.filter((t) => t.name === target.type)[0];
+      if (type?.features?.rollback) {
+        return true;
       }
-      return false;
-    },
-    [projectId, publishTypes]
-  );
+    }
+    return false;
+  };
 
   const isPullSupported = useMemo(() => {
-    if (selectedTarget) {
-      const type = publishTypes?.find((t) => t.name === selectedTarget.type);
+    return !!selectedBots.find((bot) => {
+      const publishTypes = botPublishTypesList.find((types) => types.projectId === bot.id)?.publishTypes;
+      const type = publishTypes?.find(
+        (t) => t.name === bot.publishTargets?.find((target) => target.name === bot.publishTarget)?.type
+      );
       if (type?.features?.pull) {
```

Comment:
you can just do return !!type?.features?.pull. Remove the return false after this statement